src/sign-in-xplat/MainPage.xaml.cs

```diff
@@ -0,0 +1,131 @@
+            IAccount firstAccount = accounts.FirstOrDefault();
+
+            try
+            {
+                await PublicClientApp.RemoveAsync(firstAccount);
+
+                TokenLabel.Text = "User has signed-out";
+            }
+            catch (MsalException ex)
+            {
+                TokenLabel.Text = $"Error signing-out user: {ex.Message}";
+            }
+
+            SignInButton.IsVisible = true;
+            SignOutButton.IsVisible = false;
+        }
+
+        /// <summary>
+        /// Signs in the user and obtains an Access token for MS Graph
+        /// </summary>
+        /// <param name="scopes"></param>
+        /// <returns> Access Token</returns>
+        private static async Task<string> SignInUserAndGetTokenUsingMSAL(string[] scopes)
+        {
+            // Initialize the MSAL library by building a public client application
+            PublicClientApp = PublicClientApplicationBuilder.Create(ClientId)
+                .WithAuthority(Authority)
+                .WithRedirectUri($"https://login.microsoftonline.com/common/oauth2/nativeclient")
+                .WithLogging((level, message, containsPii) =>
+                {
+                    Debug.WriteLine($"MSAL: {level} {message} ");
+                }, LogLevel.Warning, enablePiiLogging: false, enableDefaultPlatformLogging: true)
+                .Build();
+
+            IEnumerable<IAccount> accounts = await PublicClientApp.GetAccountsAsync().ConfigureAwait(false);
+            IAccount firstAccount = accounts.FirstOrDefault();
+
+            try
+            {
+                authResult = await PublicClientApp.AcquireTokenSilent(scopes, firstAccount)
+                                                  .ExecuteAsync();
+            }
+            catch (MsalUiRequiredException ex)
+            {
+                // A MsalUiRequiredException happened on AcquireTokenSilentAsync. This indicates you need to call AcquireTokenAsync to acquire a token
+                Debug.WriteLine($"MsalUiRequiredException: {ex.Message}");
+
+                authResult = await PublicClientApp.AcquireTokenInteractive(scopes)
+                                                  .ExecuteAsync()
+                                                  .ConfigureAwait(false);
+
+            }
+            return authResult.AccessToken;
+        }
+
+        /// <summary>
+        /// Sign in user using MSAL and obtain a token for MS Graph
+        /// </summary>
+        /// <returns>GraphServiceClient</returns>
+        private async static Task<GraphServiceClient> SignInAndInitializeGraphServiceClient(string[] scopes)
+        {
+            GraphServiceClient graphClient = new GraphServiceClient(MSGraphURL,
+                new DelegateAuthenticationProvider(async (requestMessage) =>
+                {
+                    requestMessage.Headers.Authorization = new AuthenticationHeaderValue("bearer", await SignInUserAndGetTokenUsingMSAL(scopes));

```

Comment:
`PuiblicClientApp`'s lifecycle, I would image, should be scoped closer to the lifecycle of the app.  In the way that you're presenting it here, you'll always need to sign in and will never leverage the token cache from what I can tell.  Maybe you move the `PublicClientApp` lifecycle out to app scope, then a call like this line would be cheaper as it would leverage the token cache by default and only fall back to interactive when necessary (instead of always).

The other issue with `new`ing up a `PublicClientApp` is that the HTTP request to the OIDC metadata endpoint would happen each time, when in reality that http client response would also be cached.

Maybe I'm wrong with all of this and it's somehow caching it all, but I'm not quite following how that would be.

Reply:
I'm sure we can find some way of moving out the PublicClientApp `building` from here and make this more performant.

Please let's confirm we'd rather prefer performance over exposing/in-lining as a way to help the user see (without jumping context) what code they are responsible for.

Reply:
I think we want to show proper lifecycle usage of the msal library.  In other languages we "new up the client" before we use it, external to where we use it.  Because that's the lifecycle of the client.  In the webapp version for python, you don't manage the client itself, but instead you need to manage the lifecycle of the token AND http cache, which you pass into the client.  Here, I'd imagine we'd just need to handle the lifecycle of the client.

If for some reason you feel it's best to leave it inline, then you'll want a clear comment above it indicating that this isn't demonstrating proper lifecycle management, I would think.

Reply:
thanks for the feedback @ckittel, we inlined it and we are now making a null check as a way to improve this.

done | addressed from [66539](https://github.com/Azure-Samples/ms-identity-docs-code-csharp/pull/11/commits/66539567652a2db7410b798a37792d8b828e4d84)

Reply:
Lifecycle changes look good in #11, thanks.